### PR TITLE
Adding features for openshift/OVA for multinodes in operator.

### PR DESCRIPTION
**Pull request by vijayrap**
Added changes for OVA?openshift on multinode , also updated to change to project the status as "Functional" and "Non-Functional" instead of Active/NotActive.

### CODE REVIEW

File: pkg/controller/contrailmonitor/contrailmonitor_controller.go [hunk 1]
```diff
@@ -316,85 +352,132 @@ func (r *ReconcileContrailmonitor) Reconcile(request reconcile.Request) (reconci
...
+	config, err := r.getConfig(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if err = r.kubernetes.Owner(instance).EnsureOwns(config); err != nil {
+		return reconcile.Result{}, err
+	}
+	if err != nil {
+		return reconcile.Result{}, err
+	}
```
Comment (rkoverzin) [inline pkg/controller/contrailmonitor/contrailmonitor_controller.go:416]: Redundant to line 411.

**Review by rkoverzin — CHANGES_REQUESTED** [review]
There are a few redundant err checks that can be removed.  I have some question on testing Active != nil implying that it is set to true.  If that is adequate then that's fine.  I had a concern that it could be not nil but set to false.

File: pkg/controller/contrailmonitor/contrailmonitor_controller.go [hunk 2]
```diff
@@ -232,12 +265,13 @@ func (r *ReconcileContrailmonitor) Reconcile(request reconcile.Request) (reconci
...
+	rablist, _ := r.getRabbitlist()
+	serInsrabbitmq := &contrailv1alpha1.Contrailstatusmonitor{ObjectMeta: metav1.ObjectMeta{Name: rablist.Items[0].Name, Namespace: "contrail"}}
 	var datarabbitmq string
-	if rabbitmq.Status.Active == nil {
-		datarabbitmq = "NotActive"
+	if rablist.Items[0].Status.Active == nil {
+		datarabbitmq = "Non-Functional"
 	} else {
-		datarabbitmq = "Active"
+		datarabbitmq = "Functional"
```
[the remark this answers is not in the document]
Comment (vijayrap) [reply pkg/controller/contrailmonitor/contrailmonitor_controller.go:262]: The work is written like that, I tried to print the whole rest API structure when it fails as well as when it passed, AS per the work flow it is populating the value to "true" when it passed or else it dosen't polulate any thing(it is blank). One more thing is that the variable declared as pointer, If we  use pointer to verify , GCP build will throw  exception mentioning Pointer can't be used, so this is the approriate approch to check the values

[the remark this answers is not in the document]
Comment (vijayrap) [reply pkg/controller/contrailmonitor/contrailmonitor_controller.go:262]: Sample o/p on various o/p.

vraparla@vraparla-mbp env % kubectl get contrailstatusmonitor -n contrailNAME         STATUS           ERRORNOTES
cassandra1   Functional       
keystone     Non-Functional   
memcached    Functional       
postgres     Functional       
rabbitmq1    Functional       
zookeeper1   Functional       
vraparla@vraparla-mbp env % kubectl get contrailstatusmonitor -n contrail
NAME                          STATUS           ERRORNOTES
cassandra1                    Functional       
contrail-analytics-api-one    Functional       
contrail-api-one              Non-Functional   
contrail-collector-one        Functional       
contrail-device-manager-one   backup           
contrail-schema-one           initializing     
contrail-svc-monitor-one      initializing     
keystone                      Non-Functional   
memcached                     Functional       
postgres                      Functional       
rabbitmq1                     Functional       
zookeeper1                    Functional       
vraparla@vraparla-mbp env % kubectl get pods -n contrail                 
NAME                                          READY   STATUS             RESTARTS   AGE
cassandra1-cassandra-statefulset-0            1/1     Running            0          2m11s
config1-config-statefulset-0                  7/10    CrashLoopBackOff   7          78s
control1-control-statefulset-0                0/4     Init:0/1           0          5s
keystone-bootstrap-job-pl66s                  0/1     Completed          0          90s
keystone-keystone-statefulset-0               0/1     Running            0          90s
memcached-deployment-58d545b4c-wp6p2          1/1     Running            0          2m11s
postgres-statefulset-0                        1/1     Running            0          2m10s
provmanager1-provisionmanager-statefulset-0   0/1     Init:0/1           0          5s
rabbitmq1-rabbitmq-statefulset-0              1/1     Running            0          2m11s
webui1-webui-statefulset-0                    0/3     Init:0/1           0          5s
zookeeper1-zookeeper-statefulset-0            1/1     Running            0          2m11s
vraparla@vraparla-mbp env % kubectl get contrailstatusmonitor -n contrail
NAME                          STATUS           ERRORNOTES
cassandra1                    Functional       
contrail-analytics-api-one    Functional       
contrail-api-one              Non-Functional   
contrail-collector-one        Functional       
contrail-device-manager-one   backup           
contrail-schema-one           initializing     
contrail-svc-monitor-one      initializing     
keystone                      Non-Functional   
memcached                     Functional       
postgres                      Functional       
provmanager1                  Functional       
rabbitmq1                     Functional       
redis-one                     Non-Functional   
webuijob-one                  Non-Functional   
webuiweb-one                  Non-Functional   
zookeeper1                    Functional

File: pkg/controller/contrailmonitor/contrailmonitor_controller.go [hunk 3]
```diff
@@ -153,71 +166,91 @@ func (r *ReconcileContrailmonitor) Reconcile(request reconcile.Request) (reconci
...
+		keystone, err := r.getKeystone(instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if err = r.kubernetes.Owner(instance).EnsureOwns(keystone); err != nil {
+			return reconcile.Result{}, err
+		}
+		if err != nil {
+			return reconcile.Result{}, err
+		}
```
[the remark this answers is not in the document]
Comment (vijayrap) [reply pkg/controller/contrailmonitor/contrailmonitor_controller.go]: Updated, thank you

File: pkg/controller/contrailmonitor/contrailmonitor_controller.go [hunk 4]
```diff
@@ -153,71 +166,91 @@ func (r *ReconcileContrailmonitor) Reconcile(request reconcile.Request) (reconci
...
+		memcached, err := r.getMemcached(instance)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		if err = r.kubernetes.Owner(instance).EnsureOwns(memcached); err != nil {
+			return reconcile.Result{}, err
+		}
+		if err != nil {
+			return reconcile.Result{}, err
+		}
```
[the remark this answers is not in the document]
Comment (vijayrap) [reply pkg/controller/contrailmonitor/contrailmonitor_controller.go]: Updated, thank you.

File: pkg/controller/contrailmonitor/contrailmonitor_controller.go [hunk 5]
```diff
@@ -259,14 +293,15 @@ func (r *ReconcileContrailmonitor) Reconcile(request reconcile.Request) (reconci
...
 
 	var datazookeeper string
-	if zookeeper.Status.Active == nil || !zookeeperActive {
-		datazookeeper = "NotActive"
+	if zoolist.Items[0].Status.Active == nil || !zookeeperActive {
+		datazookeeper = "Non-Functional"
 	} else {
-		datazookeeper = "Active"
+		datazookeeper = "Functional"
 	}
```
[the remark this answers is not in the document]
Comment (vijayrap) [reply pkg/controller/contrailmonitor/contrailmonitor_controller.go:290]: The work is written like that, I tried to print the whole rest API structure when it fails as well as when it passed, AS per the work flow it is populating the value to "true" when it passed or else it dosen't polulate any thing(it is blank). One more thing is that the variable declared as pointer, If we use pointer to verify , GCP build will throw exception mentioning Pointer can't be used, so this is the approriate approch to check the values

File: pkg/controller/contrailmonitor/contrailmonitor_controller.go [hunk 6]
```diff
@@ -288,14 +323,15 @@ func (r *ReconcileContrailmonitor) Reconcile(request reconcile.Request) (reconci
...
 
 	var datacassandra string
-	if cassandra.Status.Active == nil || !cassandraActive {
-		datacassandra = "NotActive"
+	if casslist.Items[0].Status.Active == nil || !cassandraActive {
+		datacassandra = "Non-Functional"
 	} else {
-		datacassandra = "Active"
+		datacassandra = "Functional"
 	}
```
[the remark this answers is not in the document]
Comment (vijayrap) [reply pkg/controller/contrailmonitor/contrailmonitor_controller.go:317]: The work is written like that, I tried to print the whole rest API structure when it fails as well as when it passed, AS per the work flow it is populating the value to "true" when it passed or else it dosen't polulate any thing(it is blank). One more thing is that the variable declared as pointer, If we use pointer to verify , GCP build will throw exception mentioning Pointer can't be used, so this is the approriate approach to check the values

[the remark this answers is not in the document]
Comment (vijayrap) [reply pkg/controller/contrailmonitor/contrailmonitor_controller.go:317]: vraparla@vraparla-mbp env % kubectl get contrailstatusmonitor -n contrailNAME         STATUS           ERRORNOTES
cassandra1   Functional       
keystone     Non-Functional   
memcached    Functional       
postgres     Functional       
rabbitmq1    Functional       
zookeeper1   Functional       
vraparla@vraparla-mbp env % kubectl get contrailstatusmonitor -n contrail
NAME                          STATUS           ERRORNOTES
cassandra1                    Functional       
contrail-analytics-api-one    Functional       
contrail-api-one              Non-Functional   
contrail-collector-one        Functional       
contrail-device-manager-one   backup           
contrail-schema-one           initializing     
contrail-svc-monitor-one      initializing     
keystone                      Non-Functional   
memcached                     Functional       
postgres                      Functional       
rabbitmq1                     Functional       
zookeeper1                    Functional       
vraparla@vraparla-mbp env % kubectl get pods -n contrail                 
NAME                                          READY   STATUS             RESTARTS   AGE
cassandra1-cassandra-statefulset-0            1/1     Running            0          2m11s
config1-config-statefulset-0                  7/10    CrashLoopBackOff   7          78s
control1-control-statefulset-0                0/4     Init:0/1           0          5s
keystone-bootstrap-job-pl66s                  0/1     Completed          0          90s
keystone-keystone-statefulset-0               0/1     Running            0          90s
memcached-deployment-58d545b4c-wp6p2          1/1     Running            0          2m11s
postgres-statefulset-0                        1/1     Running            0          2m10s
provmanager1-provisionmanager-statefulset-0   0/1     Init:0/1           0          5s
rabbitmq1-rabbitmq-statefulset-0              1/1     Running            0          2m11s
webui1-webui-statefulset-0                    0/3     Init:0/1           0          5s
zookeeper1-zookeeper-statefulset-0            1/1     Running            0          2m11s
vraparla@vraparla-mbp env % kubectl get contrailstatusmonitor -n contrail
NAME                          STATUS           ERRORNOTES
cassandra1                    Functional       
contrail-analytics-api-one    Functional       
contrail-api-one              Non-Functional   
contrail-collector-one        Functional       
contrail-device-manager-one   backup           
contrail-schema-one           initializing     
contrail-svc-monitor-one      initializing     
keystone                      Non-Functional   
memcached                     Functional       
postgres                      Functional       
provmanager1                  Functional       
rabbitmq1                     Functional       
redis-one                     Non-Functional   
webuijob-one                  Non-Functional   
webuiweb-one                  Non-Functional   
zookeeper1                    Functional

**Review by vijayrap — COMMENTED** [review]
vraparla@vraparla-mbp env % kubectl get pods -n contrail
NAME                                          READY   STATUS      RESTARTS   AGE
cassandra1-cassandra-statefulset-0            1/1     Running     0          7m9s
config1-config-statefulset-0                  10/10   Running     11         6m16s
control1-control-statefulset-0                4/4     Running     0          5m3s
keystone-bootstrap-job-pl66s                  0/1     Completed   0          6m28s
keystone-keystone-statefulset-0               1/1     Running     0          6m28s
memcached-deployment-58d545b4c-wp6p2          1/1     Running     0          7m9s
postgres-statefulset-0                        1/1     Running     0          7m8s
provmanager1-provisionmanager-statefulset-0   1/1     Running     1          5m3s
rabbitmq1-rabbitmq-statefulset-0              1/1     Running     0          7m9s
webui1-webui-statefulset-0                    3/3     Running     0          5m3s
zookeeper1-zookeeper-statefulset-0            1/1     Running     0          7m9s
vraparla@vraparla-mbp env % kubectl get contrailstatusmonitor -n contrail
NAME                             STATUS       ERRORNOTES
cassandra1                       Functional   
contrail-analytics-api-one       Functional   
contrail-api-one                 Functional   
contrail-collector-one           Functional   
contrail-device-manager-one      Functional   
contrail-schema-one              Functional   
contrail-svc-monitor-one         Functional   
control-collector-one            Up           
control-database-cassandra-one   Up           
control-database-rabbitmq-one    Up           
keystone                         Functional   
memcached                        Functional   
postgres                         Functional   
provmanager1                     Functional   
rabbitmq1                        Functional   
redis-one                        Functional   
webuijob-one                     Functional   
webuiweb-one                     Functional   
zookeeper1                       Functional   
vraparla@vraparla-mbp env %

**Review by rkoverzin — APPROVED** [review]

File: pkg/controller/contrailmonitor/contrailmonitor_controller.go (hunk 1 above)
Comment (vijayrap) [reply pkg/controller/contrailmonitor/contrailmonitor_controller.go]: Updated

**Review by vvaitla-juni — APPROVED** [review]
Ray has approved so i'm approving it

**Comment by vijayrap** [comment]
recheck

**Comment by vijayrap** [comment]
recheck ha

**Comment by vijayrap** [comment]
recheck ha

**Pull request merged**